tools/ci_connector_ops/ci_connector_ops/pipelines/publish.py

```diff
@@ -0,0 +1,206 @@
+            return f"{self.context.docker_image_from_metadata}-dev.{self.context.git_revision[:10]}"
+        else:
+            return self.context.docker_image_from_metadata
+
+
+class CheckConnectorImageDoesNotExist(PublishStep):
+    title = "Check if the connector docker image does not exist on the registry."
+
+    async def _run(self) -> StepResult:
+        manifest_inspect = (
+            environments.with_docker_cli(self.context)
+            .with_env_variable("CACHEBUSTER", str(uuid.uuid4()))
+            .with_exec(["docker", "manifest", "inspect", self.docker_image_name])
+        )
+        manifest_inspect_stderr = await with_stderr(manifest_inspect)
+        manifest_inspect_stdout = await with_stdout(manifest_inspect)
+
+        if "no such manifest" in manifest_inspect_stderr:
+            return StepResult(self, status=StepStatus.SUCCESS, stdout=f"No manifest found for {self.context.docker_image_from_metadata}.")
+        else:
+            try:
+                manifests = json.loads(manifest_inspect_stdout.replace("\n", "")).get("manifests", [])
+                available_platforms = {f"{manifest['platform']['os']}/{manifest['platform']['architecture']}" for manifest in manifests}
+                if available_platforms.issubset(set(builds.BUILD_PLATFORMS)):
+                    return StepResult(self, status=StepStatus.FAILURE, stderr=f"{self.context.docker_image_from_metadata} already exists.")
+                else:
+                    return StepResult(
+                        self, status=StepStatus.SUCCESS, stdout=f"No all manifests found for {self.context.docker_image_from_metadata}."
+                    )
+            except json.JSONDecodeError:
+                return StepResult(self, status=StepStatus.FAILURE, stderr=manifest_inspect_stderr, stdout=manifest_inspect_stdout)
+
+
+class BuildConnectorForPublish(Step):
+
+    title = "Build connector for publish"
+
+    async def _run(self) -> Tuple[StepResult, List[Container]]:
+        build_connector_results_per_platform = await builds.run_connector_build(self.context)
+        build_connector_results = [build_result for build_result, _ in build_connector_results_per_platform.values()]
+        built_connectors = [built_connector for _, built_connector in build_connector_results_per_platform.values()]
+
+        if not all([build_result.status is StepStatus.SUCCESS for build_result in build_connector_results]):
+            return StepResult(self, status=StepStatus.FAILURE), []
+
+        return StepResult(self, status=StepStatus.SUCCESS), built_connectors
+
+
+class PushConnectorImageToRegistry(PublishStep):
+    title = "Push connector image to registry"
+
+    async def _run(self, built_containers_per_platform: List[Container]) -> StepResult:
+        try:
+            image_ref = await built_containers_per_platform[0].publish(
+                f"docker.io/{self.docker_image_name}", platform_variants=built_containers_per_platform[1:]
+            )
```

Comment:
How does this work? I thought you couldn't have multiple arch variants of the same connector locally.  Are you renaming them?

Reply:
In this situation the connector does not have actual images in the user's docker host. These are dagger containers objects. Container objects are lazily evaluated. The `publish` function actually perform the `docker build`.
But it's dagger directly interacting with buildkit to compose the image, it's never materialized as a real image on the host. It's why I implemented [LoadContainerToLocalDockerHost](https://github.com/airbytehq/airbyte/blob/3dd75291fa51b4e1608e6d84d58986039c748e88/tools/ci_connector_ops/ci_connector_ops/pipelines/builds/common.py#L24) to make these connector images built with dagger available on the user's docker host. But it happens only on `build` not on `publish` command.

TL;DR: When calling `publish` on a Dagger container you can pass other platform variants containers. Dagger is internally responsible for the `buildx, manifest, push etc.` steps we originally have in `manage.sh`.

Reply:
This is all very cool and much cleaner than buildx